xla/service/gpu/gpu_windowed_einsum_handler.cc

```diff
@@ -94,54 +94,55 @@ absl::Status ShiftDequantizationF8(const HloComputation* comp,
         binaries[k]->shape().element_type() != F16 &&
         binaries[k]->shape().element_type() != F32) {
       VLOG(5) << "Unsupported types.";
-      return absl::OkStatus();
+      return false;
     }
   }
 
   // The FP8 scaling operands must be scalars.
   if (!ShapeUtil::IsScalar(scales[0]->shape()) ||
       !ShapeUtil::IsScalar(scales[1]->shape())) {
     VLOG(5) << "Scaling factors must be scalars.";
-    return absl::OkStatus();
+    return false;
   }
 
-  // Identify the dot and collective-permute or dynamic-slice instructions in
-  // the all-gather or reduce-scatter patterns in while's body.
+  // Identify the dot, get-tuple-element and collective-permute or dynamic-slice
+  // instructions in the all-gather or reduce-scatter patterns in while's body.
   HloComputation* while_body = while_instr->while_body();
   HloComputation* while_condition = while_instr->while_condition();
   HloInstruction* while_root = while_body->root_instruction();
-  std::array<HloInstruction*, 2> dots, dyn_slices{nullptr, nullptr},
+  std::array<HloInstruction*, 2> dots, gtes, dyn_slices{nullptr, nullptr},
       coll_perms{nullptr, nullptr};
-  if (Match(
-          while_root,
-          m::Tuple(m::CollectivePermute(
-                       &coll_perms[1], m::CollectivePermute(
-                                           &coll_perms[0], m::Op().Is(gte[0]))),
-                   m::Op().Is(gte[1]),
-                   m::DynamicUpdateSlice(
-                       m::DynamicUpdateSlice().WithOperand(
-                           1, m::Dot(&dots[0], m::Op().Is(gte[0]),
-                                     m::Op().Is(gte[1]))),
-                       m::Dot(&dots[1], m::Op(), m::Op().Is(gte[1])), m::Op(),
-                       m::Op(), m::Op()),
-                   m::Op(), m::Op()))) {
+  if (Match(while_root,
```

Comment:
I don't really understand what you are matching here exactly but it does look quite specific. Does this work on a wider range of models, e.g. will this apply if I run this on maxtext Llama 2?

Reply:
This matches the structure of the while loops of all-gather or reduce-scatter windowed einsum patterns. The patterns are generated by the SPMD dot handler and should not depend on the model.

Reply:
One example here: I find it strange that the pattern matches a specific size tuple. If you rely on the SPMD partitioner to generate this specific pattern that seems overly fragile. What if the partitioner changes? Do the transformations here depend on the number of values passed through the while loop?